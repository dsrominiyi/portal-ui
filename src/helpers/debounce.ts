/**
 * Prevents an action from being called more that once every `wait` milliseconds
 * The initial triggering of the action will be delayed by `wait` milliseconds if `immediate` = false
 *
 * e.g. <button onClick={event => debounce(handleClick, 200)(event)} />
 */

let timeout: number | null;

const debounce = (action: (...args: any[]) => void, wait = 300, immediate = true) => (
  ...args: any[]
) => {
  const later = () => {
    timeout = null;
    if (!immediate) {
      action(...args);
    }
  };
  const callNow = immediate && !timeout;

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = (setTimeout as Window['setTimeout'])(later, wait);

  if (callNow) {
    action(...args);
  }
};

export default debounce;
