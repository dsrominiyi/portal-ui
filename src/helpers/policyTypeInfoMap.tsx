import React from 'react';

import PrivateCarIcon from '../assets/svg/icon-pt-private-car.svg';
import HouseholdIcon from '../assets/svg/icon-pt-household.svg';
import TravelIcon from '../assets/svg/icon-pt-travel.svg';
import CaravanIcon from '../assets/svg/icon-pt-caravan.svg';
import TruckIcon from '../assets/svg/icon-pt-truck.svg';
import CommercialCombinedIcon from '../assets/svg/icon-pt-commercial-combined.svg';
import PublicLiabilityIcon from '../assets/svg/icon-pt-public-liability.svg';
import CaterersIcon from '../assets/svg/icon-pt-caterers.svg';

export enum PolicyType {
  PrivateCar = 'pc',
  Household = 'hh',
  Travel = 'tv',
  Caravan = 'cn',
  Truck = 'cvt',
  CommercialCombined = 'cc',
  PublicLiability = 'pl',
  Caterers = 'ct'
}

export const policyTypeInfoMap = {
  [PolicyType.PrivateCar]: { icon: <PrivateCarIcon />, displayName: 'Private Car' },
  [PolicyType.Household]: { icon: <HouseholdIcon />, displayName: 'Household' },
  [PolicyType.Travel]: { icon: <TravelIcon />, displayName: 'Travel' },
  [PolicyType.Caravan]: { icon: <CaravanIcon />, displayName: 'Caravan' },
  [PolicyType.Truck]: { icon: <TruckIcon />, displayName: 'Truck' },
  [PolicyType.CommercialCombined]: {
    icon: <CommercialCombinedIcon />,
    displayName: 'Commercial Combined'
  },
  [PolicyType.PublicLiability]: {
    icon: <PublicLiabilityIcon />,
    displayName: 'Public Liability'
  },
  [PolicyType.Caterers]: { icon: <CaterersIcon />, displayName: 'Caterers' }
};
