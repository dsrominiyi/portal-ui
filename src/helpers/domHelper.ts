import debounce from './debounce';

export const preventBodyScroll = () => {
  const { scrollY } = window;
  document.body.style.position = 'fixed';
  document.body.style.top = document.body.style.top || `-${scrollY}px`;
};

export const allowBodyScroll = () => {
  const scrollY = document.body.style.top;
  document.body.style.position = '';
  document.body.style.top = '';
  return scrollY && window.scrollTo(0, -parseInt(scrollY, 10));
};

export const triggerOnEnter = (action: () => void, debounceWait?: number) => (
  event: React.KeyboardEvent<Element | HTMLElement>
) => {
  if (event.key === 'Enter') {
    debounce(action, debounceWait)();
  }
};
