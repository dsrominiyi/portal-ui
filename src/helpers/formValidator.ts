import { utc as moment } from 'moment';

import debounce from './debounce';
import { dateFieldValuesToDateString } from './utilFuncs';

/**
 * The formValidator helper functions supports a few different types of validation, these are
 * 'required', 'minLength', 'maxLength' and 'regex', where supplied values are simply used to
 * validate, additionally `predefinedRules` are made available, being made up of a range of rule
 * parts, these can then be imported for use in a form, and adjusted as individual needs require.
 */

export interface ValidationResult {
  isValid: boolean;
  message?: string;
}

export interface ValidationError {
  showError: boolean;
  message?: string;
}

export interface FormData {
  [formFieldKey: string]: FieldValue;
}

export enum RuleType {
  postcode = 'postcode'
}

export interface ValidationRule {
  fieldLabel: string;
  minLength?: number;
  maxLength?: number;
  required?: boolean;
  regex?: string | RegExp;
  isDate?: boolean;
  invalidValues?: string[];
  errorMessage?: string;
}

export interface ValidationRules {
  [formFieldKey: string]: ValidationRule;
}

export interface BatchValidationResult {
  [formFieldKey: string]: ValidationResult;
}

export interface FormErrors {
  [formFieldKey: string]: ValidationError;
}

export interface ValidationData {
  [formFieldKey: string]: ValidationResult;
}

export type ValidationStateArray = [
  BatchValidationResult,
  React.Dispatch<React.SetStateAction<BatchValidationResult>>
];

const isDateFieldValues = (value: FieldValue): value is DateFieldValues => {
  const objKeys = typeof value === 'object' && Object.keys(value as DateFieldValues);
  return objKeys && objKeys.includes('dd') && objKeys.includes('mm') && objKeys.includes('yyyy');
};

export const predefinedRules: ValidationRules = {
  postcode: {
    fieldLabel: 'postcode',
    minLength: 5,
    maxLength: 8,
    regex: /^[A-Z]{1,2}[0-9]{1,2}[A-Z]?(\s*[0-9][A-Z]{1,2})?$/,
    errorMessage: 'You must enter a valid UK postcode'
  },
  emailAddress: {
    fieldLabel: 'email address',
    maxLength: 254,
    regex: /^[^@]+@[^.]+\..+$/,
    errorMessage: 'You must enter a valid email address',
    required: true
  },
  phoneNumber: {
    fieldLabel: 'phone number',
    minLength: 7,
    maxLength: 50,
    regex: /^\+?[\d\s]+$/,
    errorMessage: 'You must enter a valid phone number',
    required: true
  },
  individualName: {
    fieldLabel: 'name',
    regex: /^[a-zA-Z-]+$/,
    minLength: 2,
    maxLength: 50
  },
  entityName: {
    fieldLabel: 'name',
    regex: /^[\w\s-]+$/,
    minLength: 2,
    maxLength: 50
  },
  registeredCoNumber: {
    fieldLabel: 'registration number',
    regex: /^([a-zA-Z]{2}[0-9]{6}|[0-9]{8})$/,
    errorMessage: 'You must enter a valid Companies House CRN',
    required: true
  }
};

export const validate = (rule: ValidationRule, value: FieldValue = ''): ValidationResult => {
  const {
    minLength,
    maxLength,
    required,
    regex,
    isDate,
    invalidValues,
    fieldLabel,
    errorMessage
  } = rule;

  const isBoolean = typeof value === 'boolean';
  const isArray = Array.isArray(value);
  const isEmptyArray = isArray && !(value as unknown[]).length;
  const nonRequiredAndEmpty = (!required && !value) || (!required && isEmptyArray);
  const requiredAndEmpty = (required && !value) || (required && isEmptyArray);

  if (isBoolean || nonRequiredAndEmpty) {
    return { isValid: true };
  }
  if (requiredAndEmpty) {
    return {
      isValid: false,
      message: errorMessage || `Please enter ${fieldLabel}`
    };
  }

  if (isDate) {
    const dateString = isDateFieldValues(value)
      ? dateFieldValuesToDateString(value)
      : (value as string);

    return (!dateString && !required) || moment(dateString, 'YYYY-MM-DD').isValid()
      ? { isValid: true }
      : {
          isValid: false,
          message: errorMessage || `${fieldLabel} should be a valid date`
        };
  }

  const stringValue = value as string;

  if (invalidValues && invalidValues.includes(stringValue)) {
    return {
      isValid: false,
      message:
        errorMessage || `${fieldLabel} cannot be one of the following: ${invalidValues.join(', ')}`
    };
  }
  if (minLength && minLength > stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a minimum length of ${minLength} characters`
    };
  }
  if (maxLength && maxLength < stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a maximum length of ${maxLength} characters`
    };
  }
  if (regex && !RegExp(regex).test(stringValue)) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} does not match the required format`
    };
  }

  return { isValid: true };
};

export const validateAll = (rules: ValidationRules, data: FormData): BatchValidationResult => {
  return Object.entries(data).reduce(
    (validationData, [formFieldKey, value]) => ({
      ...validationData,
      [formFieldKey]: validate(rules[formFieldKey], value)
    }),
    {}
  );
};

export const validateForm = (
  rules: ValidationRules,
  data: FormData,
  setValidationState: React.Dispatch<React.SetStateAction<BatchValidationResult>>
) => {
  const validationActions = () => {
    const validationResult = validateAll(rules, data);
    setValidationState(validationResult);
  };

  debounce(validationActions, 250, false)();
};

export const isFormValid = (validatedFields: BatchValidationResult) => {
  if (Object.keys(validatedFields).length === 0) {
    return false;
  }
  let allFieldsValid = true;

  for (const validatedField of Object.values(validatedFields)) {
    if (!validatedField.isValid) {
      allFieldsValid = false;
      break;
    }
  }

  return allFieldsValid;
};

export const buildErrors = (
  validationState: BatchValidationResult,
  submissionAttempted: boolean
): FormErrors => {
  return Object.entries(validationState).reduce(
    (errors, [formFieldKey, result]) => ({
      ...errors,
      [formFieldKey]: {
        showError: submissionAttempted && !result.isValid,
        message: result.message
      }
    }),
    {}
  );
};
