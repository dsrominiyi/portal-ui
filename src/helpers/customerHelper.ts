import { dateFieldValuesToDateString, dateStringToDateFieldValues } from './utilFuncs';
import { predefinedRules, ValidationRules } from './formValidator';

export const companyFormValidationRules: ValidationRules = {
  companyName: { ...predefinedRules.entityName, fieldLabel: 'company name', required: true },
  tradingName: { ...predefinedRules.entityName, fieldLabel: 'trading name' },
  typeStatus: { fieldLabel: 'company legal status', invalidValues: ['Other'], required: true },
  businessClass: { fieldLabel: 'business class', required: true },
  tradingSinceDate: { fieldLabel: 'trading since', isDate: true },
  parentCompany: { fieldLabel: 'is this company a subsidiary', required: true },
  parentCompanyName: { ...predefinedRules.entityName, fieldLabel: 'parent company name' },
  registeredCoNumber: predefinedRules.registeredCoNumber,
  contactDetails: { fieldLabel: 'contact details', required: true },
  addresses: { fieldLabel: 'addresses', required: true }
};

export const isIndividualCustomer = (customer: Customer): customer is Individual => {
  return typeof (customer as Individual).name !== 'undefined';
};

export const getIndividualFullName = ({ title, forename, middleName, surname }: IndividualName) => {
  return `${title} ${forename}${middleName && ` ${middleName}`} ${surname}`;
};

export const buildCompanyFromFormData = (data: NewCompanyData): Company => ({
  ...data,
  parentCompany: data.parentCompany || false,
  tradingSinceDate: dateFieldValuesToDateString(data.tradingSinceDate)
});

export const buildFormDataFromCompany = (data: Company): NewCompanyData => ({
  ...data,
  tradingSinceDate: data.tradingSinceDate
    ? dateStringToDateFieldValues(data.tradingSinceDate)
    : { dd: '', mm: '', yyyy: '' }
});
