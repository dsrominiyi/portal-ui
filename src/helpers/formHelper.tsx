import React, { ReactNode } from 'react';

import QuestionBox from '../components/QuestionBox';
import TextField, { Props as TextFieldProps } from '../components/TextField';
import AbiSearch, { Props as AbiSearchProps } from '../components/AbiSearch';
import Dropdown, { DropdownOption, Props as DropdownProps } from '../components/Dropdown';
import MultiChoice, { Props as MultiChoiceProps } from '../components/MultiChoice';
import DateFields, { Props as DateFieldsProps } from '../forms/common/DateFields';
import Lookup from '../components/Lookup';

import { ValidationError } from './formValidator';

import formStyles from '../forms/styles.scss';

interface InputConfig {
  type: InputType;
  label: string;
  buttonLabel?: string;
  secondaryButtonLabel?: string;
  value?: FieldValue;
  codeListId?: string;
  options?: DropdownOption[] | string[];
  attributes?: { [attr: string]: string };
  onChange?:
    | TextFieldProps['onChange']
    | AbiSearchProps['onChange']
    | DropdownProps['onChange']
    | DateFieldsProps['onChange']
    | MultiChoiceProps['onClick'];
  secondaryAction?: () => void;
}

export const buildInput = (config: InputConfig, fieldError?: ValidationError) => {
  let input: ReactNode;

  switch (config.type) {
    case InputType.Text:
      input = (
        <TextField
          key="0"
          additionalClasses={[formStyles.formTextField]}
          label={config.label}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Lookup:
      input = (
        <Lookup
          key="0"
          label={config.label}
          buttonLabel={config.buttonLabel}
          secondaryButtonLabel={config.secondaryButtonLabel}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          secondaryAction={config.secondaryAction}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.AbiSearch:
      input = (
        <AbiSearch
          key="0"
          label={config.label}
          codeListId={config.codeListId || ''}
          onChange={config.onChange as AbiSearchProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Dropdown:
      input = (
        <Dropdown
          key="0"
          label={config.label}
          options={config.options!}
          selected={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as DropdownProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.DateFields:
      input = (
        <DateFields
          key="0"
          values={config.value as DateFieldValues}
          onChange={config.onChange as DateFieldsProps['onChange']}
          testId={config.attributes && config.attributes['test-id']}
        />
      );
      break;
    case InputType.MultiChoice:
      input = (
        <MultiChoice
          key="0"
          options={config.options as string[]}
          selectedOption={config.value as string}
          onClick={config.onChange as MultiChoiceProps['onClick']}
          attributes={config.attributes}
        />
      );
      break;
  }

  return (
    <QuestionBox
      key={config.label}
      text={[InputType.DateFields, InputType.MultiChoice].includes(config.type) ? config.label : ''}
      hasError={fieldError?.showError}
      errorText={fieldError?.message}
      inputFields={[input]}
    />
  );
};
