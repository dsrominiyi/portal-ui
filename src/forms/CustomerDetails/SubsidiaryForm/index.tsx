import React, { FunctionComponent, useState, useEffect } from 'react';

import SelectCompanyTypeStatus from '../SelectCompanyTypeStatus';
import AddContactDetailsField from '../AddItemsField/AddContactDetailsField';
import AddAdditionalContactsField from '../AddItemsField/AddAdditionalContactsField';
import AddAddressesField from '../AddItemsField/AddAddressesField';
import ValidationErrorsNotification from '../../common/ValidationErrorsNotification';

import {
  companyFormValidationRules as validationRules,
  buildCompanyFromFormData,
  buildFormDataFromCompany
} from '../../../helpers/customerHelper';
import {
  BatchValidationResult,
  ValidationStateArray,
  isFormValid,
  validateForm,
  buildErrors
} from '../../../helpers/formValidator';
import { buildInput } from '../../../helpers/formHelper';
import inputConfig from '../CompanyForm/inputConfig';
import { triggerOnEnter } from '../../../helpers/domHelper';

import formStyles from '../../styles.scss';

type UseStateArray = [NewCompanyData, React.Dispatch<React.SetStateAction<NewCompanyData>>];

export const SubsidiaryForm: FunctionComponent<FormProps> = ({
  additionalClasses = [],
  initialValues,
  submissionAttempted: submissionAttemptedProp,
  onChange,
  onSubmit
}: FormProps) => {
  const [state, setState]: UseStateArray = useState(
    (initialValues && buildFormDataFromCompany(initialValues as Company)) || {
      companyName: '',
      tradingName: '',
      typeStatus: undefined,
      businessClass: '',
      tradingSinceDate: { dd: '', mm: '', yyyy: '' },
      registeredCoNumber: '',
      parentCompany: false,
      parentCompanyName: '',
      contactDetails: [],
      additionalContacts: [],
      addresses: []
    }
  );
  const [validationState, setValidationState]: ValidationStateArray = useState(
    {} as BatchValidationResult
  );
  const [submissionAttempted, setSubmissionAttempted] = useState(false);

  const {
    companyName,
    tradingName,
    typeStatus,
    businessClass,
    tradingSinceDate,
    registeredCoNumber,
    contactDetails,
    additionalContacts,
    addresses
  } = state;
  const canSubmit = isFormValid(validationState);

  useEffect(() => {
    const formData = {
      companyName,
      tradingName,
      typeStatus,
      businessClass,
      tradingSinceDate,
      registeredCoNumber,
      contactDetails,
      addresses
    };
    validateForm(validationRules, formData, setValidationState);
    onChange(buildCompanyFromFormData(state), canSubmit);
  }, [state]);

  useEffect(() => {
    onChange(buildCompanyFromFormData(state), canSubmit);
  }, [validationState]);

  const submit = () => {
    setSubmissionAttempted(true);
    return canSubmit && onSubmit && onSubmit();
  };

  const hasAttemptedToSubmit = submissionAttemptedProp || submissionAttempted;
  const formErrors = buildErrors(validationState, hasAttemptedToSubmit);

  return (
    <div
      className={[formStyles.formSection, additionalClasses].join(' ')}
      onKeyUp={triggerOnEnter(submit)}
      test-id="new-subsidiary-form"
    >
      {hasAttemptedToSubmit && !canSubmit && <ValidationErrorsNotification />}
      <h3>Company Details</h3>
      {buildInput(
        inputConfig.companyName(companyName, event => {
          setState({ ...state, companyName: event.target.value });
        }),
        formErrors.companyName
      )}
      {buildInput(
        inputConfig.tradingName(tradingName, event => {
          setState({ ...state, tradingName: event.target.value });
        }),
        formErrors.tradingName
      )}
      <SelectCompanyTypeStatus
        typeStatus={typeStatus}
        hasError={formErrors.typeStatus?.showError}
        onChange={value => setState({ ...state, typeStatus: value })}
      />
      {buildInput(
        inputConfig.businessClass(value => {
          setState({ ...state, businessClass: value as string });
        }),
        formErrors.businessClass
      )}
      {buildInput(
        inputConfig.tradingSinceDate(tradingSinceDate, values => {
          setState({ ...state, tradingSinceDate: values });
        }),
        formErrors.tradingSinceDate
      )}
      {buildInput(
        inputConfig.registeredCoNumber(registeredCoNumber, event => {
          setState({ ...state, registeredCoNumber: event.target.value });
        }),
        formErrors.registeredCoNumber
      )}
      <AddContactDetailsField
        contactDetails={contactDetails}
        hasError={formErrors.contactDetails?.showError}
        onChange={values => setState({ ...state, contactDetails: values })}
      />
      <AddAdditionalContactsField
        additionalContacts={additionalContacts}
        onChange={values => setState({ ...state, additionalContacts: values })}
      />
      <h3>Address Details</h3>
      <AddAddressesField
        customerType="Company"
        addresses={addresses}
        hasError={formErrors.addresses?.showError}
        onChange={values => setState({ ...state, addresses: values })}
      />
    </div>
  );
};

SubsidiaryForm.displayName = 'SubsidiaryForm';
export default SubsidiaryForm;
