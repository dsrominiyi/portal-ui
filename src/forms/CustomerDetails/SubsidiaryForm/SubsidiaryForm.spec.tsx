import React from 'react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import SubsidiaryForm from '.';

import { buildCompanyFromFormData } from '../../../helpers/customerHelper';
import { isFormValid, buildErrors } from '../../../helpers/formValidator';

const mockStore = createMockStore([]);

jest.mock('../../../helpers/formValidator');
jest.useFakeTimers();

let form: ShallowWrapper | ReactWrapper;
const onChange = jest.fn();
const onSubmit = jest.fn();
(buildErrors as jest.Mock).mockReturnValue({
  companyName: {},
  tradingName: {},
  typeStatus: {},
  businessClass: {},
  tradingSinceDate: {},
  registeredCoNumber: {},
  contactDetails: {},
  addresses: {}
});

describe('SubsidiaryForm', () => {
  describe('snapshots', () => {
    beforeEach(() => {
      form = shallow(<SubsidiaryForm onChange={onChange} onSubmit={onSubmit} />);
    });

    test('SubsidiaryForm matches snapshot', () => {
      expect(form).toMatchSnapshot();
    });

    test('SubsidiaryForm with initial values matches snapshot', () => {
      (form as ShallowWrapper).setProps({
        initialValues: { companyName: 'Datamatters' }
      });

      expect(form).toMatchSnapshot();
    });

    test('SubsidiaryForm without form errors matches snapshot', () => {
      (buildErrors as jest.Mock).mockReturnValueOnce({});
      form = shallow(<SubsidiaryForm onChange={onChange} onSubmit={onSubmit} />);

      expect(form).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    const hitEnter = () => {
      act(() => {
        (form
          .find('div')
          .at(0)
          .props() as any).onKeyUp({ key: 'Enter' });
      });
      form.update();
    };

    const render = () => {
      form = mount(
        <Provider
          store={mockStore({
            search: {
              hitResults: [],
              suggestResults: []
            }
          })}
        >
          <SubsidiaryForm onChange={onChange} onSubmit={onSubmit} />
        </Provider>
      );
    };

    const runTimers = () => {
      act(() => {
        jest.runAllTimers();
      });
    };

    beforeEach(() => {
      runTimers();
      jest.clearAllMocks();
      (isFormValid as jest.Mock) = jest.fn(() => true);
      render();
    });

    test('onSubmit is called when the Enter is pressed', () => {
      hitEnter();

      expect(onSubmit).toHaveBeenCalled();
    });

    test('onSubmit is not called when the Enter is pressed if the form is invalid', () => {
      (isFormValid as jest.Mock) = jest.fn(() => false);
      render();
      hitEnter();

      expect(onSubmit).not.toHaveBeenCalled();
    });

    test('ValidationErrorsNotification is rendered when the Enter is pressed but the form is invalid', () => {
      (isFormValid as jest.Mock) = jest.fn(() => false);
      hitEnter();

      expect(form.find('ValidationErrorsNotification').length).toBe(1);
    });

    test('onChange is called with the form state when any field is updated', () => {
      const companyName = 'Datamatters Ltd';

      act(() => {
        (form.find({ 'test-id': 'company-name' }).props() as any).onChange({
          target: { value: companyName }
        });
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName: '',
          typeStatus: undefined,
          businessClass: '',
          tradingSinceDate: { dd: '', mm: '', yyyy: '' },
          registeredCoNumber: '',
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const tradingName = 'Datamatters';

      act(() => {
        (form.find({ 'test-id': 'trading-name' }).props() as any).onChange({
          target: { value: tradingName }
        });
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus: undefined,
          businessClass: '',
          tradingSinceDate: { dd: '', mm: '', yyyy: '' },
          registeredCoNumber: '',
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const typeStatus = CompanyTypeStatus.Limited;

      act(() => {
        (form.find('SelectCompanyTypeStatus').props() as any).onChange(typeStatus);
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass: '',
          tradingSinceDate: { dd: '', mm: '', yyyy: '' },
          registeredCoNumber: '',
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const businessClass = 'Accountancy';

      act(() => {
        (form.find('AbiSearch').props() as any).onChange(businessClass);
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass,
          tradingSinceDate: { dd: '', mm: '', yyyy: '' },
          registeredCoNumber: '',
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const tradingSinceDate = { dd: '22', mm: '06', yyyy: '1980' };

      act(() => {
        (form.find('DateFields').props() as any).onChange(tradingSinceDate);
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass,
          tradingSinceDate,
          registeredCoNumber: '',
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const registeredCoNumber = '12345678';

      act(() => {
        (form.find({ 'test-id': 'registration-number' }).props() as any).onChange({
          target: { value: registeredCoNumber }
        });
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass,
          tradingSinceDate,
          registeredCoNumber,
          parentCompany: false,
          parentCompanyName: '',
          contactDetails: [],
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const contactDetails = [{}, {}] as Contact[];

      act(() => {
        (form
          .find('AddContactDetailsField')
          .at(0)
          .props() as any).onChange(contactDetails);
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass,
          tradingSinceDate,
          registeredCoNumber,
          parentCompany: false,
          parentCompanyName: '',
          contactDetails,
          additionalContacts: [],
          addresses: []
        }),
        true
      );

      form.update();
      runTimers();

      const additionalContacts = [{}, {}] as AdditionalContact[];

      act(() => {
        (form.find('AddAdditionalContactsField').props() as any).onChange(additionalContacts);
      });

      expect(onChange).toHaveBeenCalledWith(
        buildCompanyFromFormData({
          companyName,
          tradingName,
          typeStatus,
          businessClass,
          tradingSinceDate,
          registeredCoNumber,
          parentCompany: false,
          parentCompanyName: '',
          contactDetails,
          additionalContacts,
          addresses: []
        }),
        true
      );
    });
  });
});
