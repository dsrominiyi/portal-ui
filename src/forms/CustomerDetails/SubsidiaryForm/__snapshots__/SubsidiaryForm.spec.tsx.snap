// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubsidiaryForm snapshots SubsidiaryForm matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-subsidiary-form"
>
  <h3>
    Company Details
  </h3>
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "company-name",
            }
          }
          label="Company Name"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Company Name"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "trading-name",
            }
          }
          label="Trading name (optional)"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Trading name (optional)"
    text=""
  />
  <SelectCompanyTypeStatus
    onChange={[Function]}
  />
  <QuestionBox
    inputFields={
      Array [
        <AbiSearch
          attributes={
            Object {
              "test-id": "business-class",
            }
          }
          codeListId="11"
          label="Business class"
          onChange={[Function]}
        />,
      ]
    }
    key="Business class"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <DateFields
          onChange={[Function]}
          testId="trading-since-date"
          values={
            Object {
              "dd": "",
              "mm": "",
              "yyyy": "",
            }
          }
        />,
      ]
    }
    key="Trading since date (optional)"
    text="Trading since date (optional)"
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "registration-number",
            }
          }
          label="Registration number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Registration number"
    text=""
  />
  <AddContactDetailsField
    contactDetails={Array []}
    onChange={[Function]}
  />
  <AddAdditionalContactsField
    additionalContacts={Array []}
    onChange={[Function]}
  />
  <h3>
    Address Details
  </h3>
  <AddAddressesField
    addresses={Array []}
    customerType="Company"
    onChange={[Function]}
  />
</div>
`;

exports[`SubsidiaryForm snapshots SubsidiaryForm with initial values matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-subsidiary-form"
>
  <h3>
    Company Details
  </h3>
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "company-name",
            }
          }
          label="Company Name"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Company Name"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "trading-name",
            }
          }
          label="Trading name (optional)"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Trading name (optional)"
    text=""
  />
  <SelectCompanyTypeStatus
    onChange={[Function]}
  />
  <QuestionBox
    inputFields={
      Array [
        <AbiSearch
          attributes={
            Object {
              "test-id": "business-class",
            }
          }
          codeListId="11"
          label="Business class"
          onChange={[Function]}
        />,
      ]
    }
    key="Business class"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <DateFields
          onChange={[Function]}
          testId="trading-since-date"
          values={
            Object {
              "dd": "",
              "mm": "",
              "yyyy": "",
            }
          }
        />,
      ]
    }
    key="Trading since date (optional)"
    text="Trading since date (optional)"
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "registration-number",
            }
          }
          label="Registration number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Registration number"
    text=""
  />
  <AddContactDetailsField
    contactDetails={Array []}
    onChange={[Function]}
  />
  <AddAdditionalContactsField
    additionalContacts={Array []}
    onChange={[Function]}
  />
  <h3>
    Address Details
  </h3>
  <AddAddressesField
    addresses={Array []}
    customerType="Company"
    onChange={[Function]}
  />
</div>
`;

exports[`SubsidiaryForm snapshots SubsidiaryForm without form errors matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="new-subsidiary-form"
>
  <h3>
    Company Details
  </h3>
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "company-name",
            }
          }
          label="Company Name"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Company Name"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "trading-name",
            }
          }
          label="Trading name (optional)"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Trading name (optional)"
    text=""
  />
  <SelectCompanyTypeStatus
    onChange={[Function]}
  />
  <QuestionBox
    inputFields={
      Array [
        <AbiSearch
          attributes={
            Object {
              "test-id": "business-class",
            }
          }
          codeListId="11"
          label="Business class"
          onChange={[Function]}
        />,
      ]
    }
    key="Business class"
    text=""
  />
  <QuestionBox
    inputFields={
      Array [
        <DateFields
          onChange={[Function]}
          testId="trading-since-date"
          values={
            Object {
              "dd": "",
              "mm": "",
              "yyyy": "",
            }
          }
        />,
      ]
    }
    key="Trading since date (optional)"
    text="Trading since date (optional)"
  />
  <QuestionBox
    inputFields={
      Array [
        <TextField
          additionalClasses={
            Array [
              "formTextField",
            ]
          }
          attributes={
            Object {
              "test-id": "registration-number",
            }
          }
          label="Registration number"
          onChange={[Function]}
          value=""
        />,
      ]
    }
    key="Registration number"
    text=""
  />
  <AddContactDetailsField
    contactDetails={Array []}
    onChange={[Function]}
  />
  <AddAdditionalContactsField
    additionalContacts={Array []}
    onChange={[Function]}
  />
  <h3>
    Address Details
  </h3>
  <AddAddressesField
    addresses={Array []}
    customerType="Company"
    onChange={[Function]}
  />
</div>
`;
